### PR TITLE
feat(devtools): allow user to change frontmatter in multiple posts

### PR DESCRIPTION
**Pull request by smallp**
### Description
Enable user to change frontmatter in multiple posts.

### Linked Issues
#174 

### Copilot Descriptions

copilot:all

### CODE REVIEW

**Comment by WRXinYue** [comment]
My suggestion is to do this in a plugin or terminal, which can be developed and maintained independently without affecting existing development tools, rather than adding it to devtools. Adding too many non-core functions might make devtools bloated.

**Comment by YunYouJun** [comment]
Without adding additional dependencies, the added code is manageable. Maybe we can design a plug-in system for devtools in the future and get it up and running before we do that.

**Comment by YunYouJun** [comment]
cc @WRXinYue  @qtqz @guowei-gong 

Are anyone interested in verifying and refining this feature?

File: packages/devtools/src/node/api/index.ts [hunk 1]
```diff
@@ -2,9 +2,33 @@ import type { ResolvedConfig, ViteDevServer } from 'vite'
 import bodyParser from 'body-parser'
 
 import matter from 'gray-matter'
+import { JSON_SCHEMA } from 'js-yaml'
 import fs from 'fs-extra'
 
 const prefix = '/valaxy-devtools-api'
+export const userroot: { root: string } = { root: '' }
```
Comment (YunYouJun) [inline packages/devtools/src/node/api/index.ts]: All `userroot` should be `userRoot`

File: packages/devtools/src/node/index.ts [hunk 2]
```diff
@@ -4,13 +4,14 @@ import sirv from 'sirv'
...
 import type { ValaxyDevtoolsOptions } from './types'
-import { registerApi } from './api'
+import { registerApi, userroot } from './api'
 
 const NAME = 'valaxy:devtools'
 
 // import.meta.env.VITE_DEV_VALAXY_DEVTOOLS = 'true'
 
 export default function ValaxyDevtools(options: ValaxyDevtoolsOptions = {}): Plugin {
+  userroot.root = options.userroot!
```
Comment (YunYouJun) [inline packages/devtools/src/node/index.ts]: Why redefine?

**Review by YunYouJun — CHANGES_REQUESTED** [review]

File: packages/devtools/src/node/api/index.ts [hunk 3]
```diff
@@ -16,7 +40,7 @@ export function registerApi(server: ViteDevServer, _viteConfig: ResolvedConfig)
     if (req.method === 'POST') {
       const { pageData, frontmatter: newFm } = await (req as any).body
       // filePath
-      const path = pageData.path
+      const path = `${userroot.root}/pages${pageData.path}.md`
```
Comment (YunYouJun) [inline packages/devtools/src/node/api/index.ts]: `pageData.path` is a full path. You can not use it like this.

File: packages/valaxy/node/server.ts [hunk 4]
```diff
@@ -26,7 +27,7 @@ export async function createServer(
     // only enable when dev
     vitePlugins.push(
       (await import('vite-plugin-vue-devtools')).default(),
-      (await import('@valaxyjs/devtools')).default(),
+      (await import('@valaxyjs/devtools')).default({ userroot: userRoot }),
```
Comment (YunYouJun) [inline packages/valaxy/node/server.ts]: useRoot can got from `options`. You should not pass it.

**Review by YunYouJun — COMMENTED** [review]
image

This pr contains the wrong type build (js-yaml) so that it cannot pass CI. For new features, and there is no style, no comments.
Before waiting for a merge, you should make sure it passes all CIS and resolves conflicts.

File: packages/devtools/src/node/api/index.ts [hunk 5]
```diff
@@ -32,4 +56,22 @@ export function registerApi(server: ViteDevServer, _viteConfig: ResolvedConfig)
...
   })
+
+  app.use(`${prefix}/migration`, async (req, _res) => {
+    // update
+    if (req.method === 'POST') {
+      const { pageData, frontmatter } = await (req as any).body
+      // filePath
+      const worker: Promise<void>[] = []
+
+      for (const item of pageData)
```
Comment (YunYouJun) [inline packages/devtools/src/node/api/index.ts]: The meaning of pagedata here does not correspond to its meaning.

**Comment by YunYouJun** [comment]
Just merge it for later work.

File: packages/valaxy/node/server.ts (hunk 4 above)
Comment (smallp) [reply packages/valaxy/node/server.ts]: That's because I dont know how to access userRoot in options.

Comment (YunYouJun) [reply packages/valaxy/node/server.ts]: Get it from `options.userRoot` directly.

**Pull request merged**